Sources/_StringProcessing/Engine/Processor.swift

```diff
@@ -94,6 +94,23 @@ extension Processor {
     assert(currentPosition >= start)
     assert(currentPosition <= end)
   }
+
+  mutating func reset(newBounds: Range<Position>)
+  {
+    self.bounds = newBounds
+
+    self.controller = Controller(pc: 0)
+    self.currentPosition = bounds.lowerBound
+
+    self.registers.reset(bounds.upperBound)
+    self.storedCaptures = self.storedCaptures.map {_ in .init()}
```

Comment:
The perf win of reset comes from re-using the same arrays instead of allocating new ones.

```suggestion
    for idx in self.storedCaptures.indices {
      self.storedCaptures[idx] = .init()
    }
```

(@natecook1000 is there an algorithm for this?)

Reply:
No, we've had a mutating `forEach` pitched several times, but it's never been added to the stdlib or `swift-algorithms`.